Custom auth guide updates (#3998)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/authenticators-custom-authenticator/main/ios/sdksteps.md

> **Note:** The iOS code snippets assume that there's a singleton class called `DeviceSDKManager` that manages interactions with the Devices SDK. The singleton contains any required state information, such as the current SDK client, the APNs token for the current launch of the app, utility functions, and more. Your app may use a different way of interacting with the parts of the Devices SDK.

The following image shows how data flows through the Devices SDK:

<div class="full border">

![Custom authenticator flow chart](/img/authenticators/authenticators-custom-authenticator-data-flow.png)

</div>

### Enable the user to sign in

A valid user account authentication token is necessary to add a device as an authenticator. To get the token, first enable the user to sign in to their account using the [Okta mobile Swift SDK](https://github.com/okta/okta-mobile-swift) or if you're already using it, the [Okta IDX Swift SDK](https://github.com/okta/okta-idx-swift). For more information on signing users in to your app, see [Sign users in to your mobile app using the redirect model](https://developer.okta.com/docs/guides/sign-into-mobile-app-redirect/ios/main/).

> Then add the extra permissions that the Devices SDK requires to the access token. Add the following strings to the space-delimited list of scopes in the `Okta.plist` file:

Author: Might be best to pass along this feedback to Maurice.

    pod 'DeviceAuthenticator'
end
```

The source for the Okta Swift Devices SDK is in the Okta GitHub repository:
https://github.com/okta/okta-devices-swift.

There are two ways for your app to receive notifications from a custom authenticator:

- As a push notification that's delivered whether your app is closed, in the background, or in the foreground
- By requesting any queued notifications when your app is in the foreground

> Although you don't need to receive push notifications to use the Devices SDK, we suggest that you do this for the best user experience. To receive notifications add the Push Notification Capability to your app. At runtime, register your app with the notification manager, and retrieve the current APNs token. Use this token in the next step.

Reviewer: ```suggestion
Although your app doesn't need to receive push notifications to use the Devices SDK, we suggest that you do this for the best user experience. To receive notifications, add the Push Notification Capability to your app. At runtime, register your app with the notification manager, and retrieve the current APNs token. Use this token in the next step.
```

To use the device to verify the identity of a user it must be registered, or *enrolled*, with the custom authenticator.

> To enroll a device you need:

Reviewer: ```suggestion
To enroll a device, you need:
```

- An app that enables a user to sign in to their account

> - That requests the appropriate scopes. See [Enable the user to sign in](#enable-the-user-to-sign-in).

Reviewer: ```suggestion
- Request the appropriate scopes. See [Enable the user to sign in](#enable-the-user-to-sign-in).
```

- A configured and enabled custom authenticator in your Okta org
- The current APNs token if your app is registered for push notifications

> Alternatively, you can enroll the device using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/createAppAuthenticatorEnrollment).

Reviewer: ```suggestion
Alternatively, you can enroll the device by using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/createAppAuthenticatorEnrollment).
```

The function assumes a valid access token. In a production app, consider adding a function that either updates the token if it's invalid, or gets a new token using a user sign-in flow.

> Alternatively, you can update the registration token using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/updateAppAuthenticatorEnrollment).

Reviewer: ```suggestion
Alternatively, you can update the registration token by using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/updateAppAuthenticatorEnrollment).
```

}
```

> Alternatively, you can delete an enrollment using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/deleteAppAuthenticatorEnrollment).

Reviewer: ```suggestion
Alternatively, you can delete an enrollment by using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/deleteAppAuthenticatorEnrollment).
```

        ...
```

> The first two titles are the actions for the notification that requests a user approve or deny that they're trying to sign in. The third is the action for a biometric verification. Replace `YourAppName` with the name of your app, which you can read from the `CFBundleName` key of the `Info.plist` file in your main bundle.

Reviewer: ```suggestion
The first two titles are the actions for the notification that requests a user to approve or deny that they're trying to sign in. The third is the action for a biometric verification. Replace `YourAppName` with the name of your app, which you can read from the `CFBundleName` key of the `Info.plist` file in your main bundle.
```

```

The code assumes that each enrollment uses the same access token. Use the appropriate access token if your app handles multiple user accounts.

Alternatively, you can retrieve undelivered challenges using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/listAppAuthenticatorPendingPushNotificationChallenges).

## Access token management

> **Note:** To enable the JWT bearer grant type:
>  * Send a PUT request to `/apps/{appId}`. Ensure that the `grant_types` array contains the following string:
>    `urn:ietf:params:oauth:grant-type:jwt-bearer`
>  * If you use custom authorization servers, update the policy rules to update the grant type.

> The SDK communicates with an Okta server using HTTPS protocol and requires an access token for user authentication and authorization. For authentication flows and requesting access tokens, use the latest version of the [Okta OIDC SDK](https://github.com/okta/okta-oidc-android). To enroll a push authenticator, the access token needs the `okta.myAccount.appAuthenticator.manage` scope. You can also use this scope for the following operations:

Reviewer 2: should be
[Okta Mobile SDK](https://github.com/okta/okta-mobile-kotlin)

Author: @FeiChen-okta I just want to confirm that this is the correct SDK. You've linked to the Kotlin SDK here for the iOS guide.

```

The code assumes that each enrollment uses the same access token. Use the appropriate access token if your app handles multiple user accounts.

> Alternatively, you can retrieve undelivered challenges using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/listAppAuthenticatorPendingPushNotificationChallenges).

Reviewer: ```suggestion
Alternatively, you can retrieve undelivered challenges by using the [MyAccount App Authenticators API](https://developer.okta.com/docs/api/openapi/okta-myaccount/myaccount/tag/AppAuthenticator/#tag/AppAuthenticator/operation/listAppAuthenticatorPendingPushNotificationChallenges).
```

## Access token management

> **Note:** To enable the JWT bearer grant type:

> >  * Send a PUT request to `/apps/{appId}`. Ensure that the `grant_types` array contains the following string:

Reviewer 3: @jane-kuang-okta , do we have a good example how to do that? For me doesn't look obvious

>    `urn:ietf:params:oauth:grant-type:jwt-bearer`

> >  * If you use custom authorization servers, update the policy rules to update the grant type.

Reviewer 4: I know the link to the API is already in the nut facts. Should it be added in this note blurb as well?

Reviewer: Yes, I agree. It would be helpful to have a link to the `PUT /apps/${appId}/` endpoint description.

> The SDK communicates with an Okta server using HTTPS protocol and requires an access token for user authentication and authorization. For authentication flows and requesting access tokens, use the latest version of the [Okta OIDC SDK](https://github.com/okta/okta-oidc-android). To enroll a push authenticator, the access token needs the `okta.myAccount.appAuthenticator.manage` scope. You can also use this scope for the following operations:

Author: ```suggestion
The SDK communicates with an Okta server using HTTPS protocol and requires an access token for user authentication and authorization. For authentication flows and requesting access tokens, use the latest version of the [Okta Swift mobile SDK](https://github.com/okta/okta-mobile-swift). To enroll a push authenticator, the access token needs the `okta.myAccount.appAuthenticator.manage` scope. You can also use this scope for the following operations:
```

> The SDK communicates with an Okta server using HTTPS protocol and requires an access token for user authentication and authorization. For authentication flows and requesting access tokens, use the latest version of the [Okta Swift mobile SDK](https://github.com/okta/okta-mobile-swift). To enroll a push authenticator, the access token needs the `okta.myAccount.appAuthenticator.manage` scope. You can also use this scope for the following operations:

Reviewer: ```suggestion
The SDK communicates with an Okta server using the HTTPS protocol and requires an access token for user authentication and authorization. For authentication flows and access token requests, use the latest version of the [Okta Swift mobile SDK](https://github.com/okta/okta-mobile-swift). To enroll a push authenticator, the user needs to have an access token that contains the `okta.myAccount.appAuthenticator.manage` scope. You can also use this scope for the following operations:
```

* Enroll and unenroll user verification keys
* Update device token for push authenticator enrollment
* Request pending push challenges
* Enable and disable CIBA capability for push authenticator enrollment
* Delete push authenticator enrollment

>   < **Note:** Applications that use sensitive data shouldn't store or cache access tokens or refresh access tokens that contain the `okta.myAccount.appAuthenticator.manage` scope. Instead, reauthenticate the user and get a new access token.

Reviewer 4: Typo before **Note:**

>   < **Note:** Applications that use sensitive data shouldn't store or cache access tokens or refresh access tokens that contain the `okta.myAccount.appAuthenticator.manage` scope. Instead, reauthenticate the user and get a new access token.

Reviewer: ```suggestion
  > **Note:** Applications that use sensitive data shouldn't store or cache access tokens or refresh access tokens that contain the `okta.myAccount.appAuthenticator.manage` scope. Instead, reauthenticate the user and get a new access token.
```

>   High risk operations include the following:

Reviewer: Two spaces in front of this line makes it look like a subset of the previous list.

Do we need to tell the user that high risk operations need reauthentication?
For example: "The following is a list of operations that are considered high risk and require reauthentication:"

  * Enroll push authenticator
  * Enable or disable user verification for push authenticator enrollment
  * Delete push authenticator enrollment

> Other operations are low risk and may not require interactive authentication. For that reason, the Okta OIDC SDK implements the silent user reauthentication API `retrieveMaintenanceToken`. By retrieving a maintenance access token, an application can silently perform the following operations:

Reviewer: Is the `retrieveMaintenanceToken` a reauthentication API by itself? Or perhaps a method?

"For that reason, the Okta OIDC SDK provides the silent user reauthentication method, `retrieveMaintenanceToken`. This method retrieves a maintenance access token for reauthentication that allows an application to silently perform the following operations:"

> Other operations are low risk and may not require interactive authentication. For that reason, the Okta OIDC SDK implements the silent user reauthentication API `retrieveMaintenanceToken`. By retrieving a maintenance access token, an application can silently perform the following operations:

Reviewer 3: `For that reason, the Okta OIDC SDK implements...`, not OIDC SDK but Push SDK, so should be fixed to - `For that reason, the Okta Push SDK implements...`
